Refactored Subscription

cbergmiller opened the pull request:
**preliminary - do not merge yet**

I worked on the client subscription publish mechanism. There is now less _fire and forget_ but the changes affect the  symmetry between client and server a bit (internally).
The `UaClient` now has a task that cyclically issues PublishRequests. `Subscription` no longer has to call the `publish()` method.
There is much more potential for simplification in the workings of `InternalSession`, `SubscriptionService`, `InternalSubscription`, `MonitoredItemService` ...
But before that i would like to try and draw a UML diagram, it is hard for me to keep that level of complexity in my mind.
This branch needs further testing and discussion. I may try to use it for some productive systems at work for testing.

oroulet commented:
If this is work in progress start name with 'WIP:' that is the common convention

oroulet commented:
I have not looked at code yet. But you MUST send publish request after each publish response. Most server complain if they have too large requests queue

cbergmiller commented:
Thanks for the heads up. I did not change the sequencing of the publish request, it remains the same. I just moved it into an endless loop inside of an task in UaClient. This reduces complexity in my opinion.

oroulet commented on asyncua/client/ua_client.py:
```diff
@@ -406,117 +409,103 @@ def disconnect_socket(self):
+            except Exception:
+                # INFO: catching the exception here might be obsolete because we already
+                #       catch BadTimeout above. However, it's not really clear what this code
+                #       does so it stays in, doesn't seem to hurt.
```
here we catch any parsing error. On the network you cannot trust what you receive anyway. Not sure what you mean it has to do with timeout

cbergmiller replied:
This comment was added in f88dce0d99886226eb7ef5226c3ddec067a2307a back in 2016
I can replace or remove the comment. Also i am not sure if it is the right thing to send an Publish Response without acknowledgement when the Pbulish Response cannot be parsed.

oroulet commented on asyncua/client/ua_client.py:
```diff
@@ -406,117 +409,103 @@ def disconnect_socket(self):
+        This is just for maintaining symmetry.
+        This client takes care of sending PublishRequest internally.
+        """
+        pass
```
should't we make the method really send a publish request? some people may want to do it manually. Otherwise we should probably deleted it

cbergmiller replied:
That is a good point. I will make it possible to send PublishRequest via this public method again.

oroulet commented on asyncua/common/subscription.py:
```diff
@@ -89,27 +84,18 @@ def __init__(self, server, params, handler):
-        # Send a publish request so the server has one in its queue
-        # Servers should always be able to handle at least on extra publish request per subscriptions
-        await self.server.publish()
+        #self.server.publish()
```
do not leave dead commented code

oroulet commented on asyncua/server/internal_server.py:
```diff
@@ -43,7 +34,19 @@ def __init__(self, serv, cap=None):
+        isession.user = User.Admin
+    return True
+
+
```
why do I see this? this seems to be an old diff?

cbergmiller replied:
Probably because i moved `InternalSession` into a new file for easier access.

oroulet commented:
As far as I understood you are replacing some callbacks with async/await. This is a good idea

cbergmiller commented:
Thanks for your review! Yes i am trying to remove some of the callbacks - at least in cases where this leads to cleaner architecture.

oroulet commented on asyncua/client/ua_client.py:
```diff
@@ -3,14 +3,18 @@
 
 
+class ResponseParseError(ValueError):
+    """Error while parsing responses."""
```
I am quite sure there is a PArsingError or similar already defined somewhere. it should be shared between server and client. I found it in asyncua/ua/uaerrors/_base.py... the name is strange... maybe do a grep and rename...

oroulet commented on asyncua/client/ua_client.py:
```diff
@@ -422,7 +426,7 @@ def disconnect_socket(self):
         if not self._publish_task or self._publish_task.done():
             # Start the publish cycle if it is not yet running
-            self._publish_task = self.loop.create_task(self._publish())
+            self._publish_task = self.loop.create_task(self._publish_cycle())
```
publish_loop? or publish_task? publish_cycle sounds a  bit strange..

oroulet replied:
ususally a cycle is path of a loop

cbergmiller commented:
I consider this finished and would like to merge these changes.
To recap what this includes:
- Documentation and Typings for class methods for better maintainability in the future
- `UaClient` now has a `_publish_task` that waits for the publish response in an loop (this simplifies `Subscription`)
- Added support for servers returning a publish response for subsription ID 0 (meaning there are no more subscriptions) as to spec. Part 4 - Section 5.13.5
- some cleanup and fixed typos

oroulet commented on asyncua/client/ua_client.py:
```diff
@@ -279,6 +280,8 @@ def disconnect_socket(self):
     async def close_session(self, delete_subscriptions):
         self.logger.info("close_session")
+        if self._publish_task and not self._publish_task.done():
+            self._publish_task.cancel()
```
should'nt we wait for task? just wondering

oroulet commented on asyncua/server/internal_server.py:
```diff
@@ -3,38 +3,30 @@
-    Created = 0
-    Activated = 1
-    Closed = 2
+    uacrypto = False
```
that test should be removed, now we have a hard dependency on crypto
